rust/cardano-blockchain-types/src/point.rs

```diff
@@ -0,0 +1,595 @@
+    }
+
+    /// Check if a Point is the origin.
+    ///
+    /// Origin is the synthetic Origin point, and ALSO any point thats at slot zero with a
+    /// hash.
+    ///
+    /// # Examples
+    ///
+    /// ```rs
+    /// use cardano_chain_follower::Point;
+    ///
+    /// let slot = 42;
+    /// let point = Point::fuzzy(slot);
+    ///
+    /// assert!(!point.is_origin());
+    /// ```
+    #[must_use]
+    pub fn is_origin(&self) -> bool {
+        match self.0 {
+            pallas::network::miniprotocols::Point::Specific(slot, ref hash) => {
+                slot == 0 && !hash.is_empty()
+            },
+            pallas::network::miniprotocols::Point::Origin => true,
+        }
+    }
+
+    /// Check if a Point is actually unknown.
+    ///
+    /// # Examples
+    ///
+    /// ```rs
+    /// use cardano_chain_follower::Point;
+    ///
+    /// let point = Point::fuzzy(0);
+    ///
+    /// assert!(point.is_unknown());
+    /// ```
+    #[must_use]
+    pub fn is_unknown(&self) -> bool {
+        match self.0 {
+            pallas::network::miniprotocols::Point::Specific(slot, ref hash) => {
+                slot == 0 && hash.is_empty()
+            },
+            pallas::network::miniprotocols::Point::Origin => false,
+        }
+    }
+
+    /// Check if a Point is actually unknown.
+    ///
+    /// # Examples
+    ///
+    /// ```rs
+    /// use cardano_chain_follower::Point;
+    ///
+    /// let point = Point::fuzzy(0);
+    ///
+    /// assert!(point.is_unknown());
+    /// ```
+    #[must_use]
+    pub fn is_tip(&self) -> bool {
+        match self.0 {
+            pallas::network::miniprotocols::Point::Specific(slot, ref hash) => {
+                slot == u64::MAX && hash.is_empty()
+            },
+            pallas::network::miniprotocols::Point::Origin => false,
+        }
+    }
+
+    /// Compares the hash stored in the `Point` with a known hash.
+    /// It returns `true` if the hashes match and `false` otherwise. If the
+    /// provided hash is `None`, the function checks if the `Point` has an
+    /// empty hash.
+    ///
+    /// # Parameters
+    ///
+    /// * `hash` - An `Option<Hash<32>>` containing the hash to compare against. If
+    ///   `Some`, the contained hash is compared with the `Point`'s hash. If `None`, the
+    ///   function checks if the `Point`'s hash is empty.
+    ///
+    /// # Returns
+    ///
+    /// A `bool` indicating whether the hashes match.
+    ///
+    /// # Examples
+    ///
+    /// ```rs
+    /// use cardano_chain_follower::Point;
+    ///
+    /// use pallas::crypto::hash::Hash;
+    ///
+    /// let point = Point::new(42, vec![1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11]);
+    /// let hash = Some(Hash::new([1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11]));
+    /// assert!(point.cmp_hash(&hash));
+    ///
+    /// let empty_point = Point::fuzzy(42);
+    /// assert!(empty_point.cmp_hash(&None));
+    /// ```
+    #[must_use]
+    pub fn cmp_hash(&self, hash: &Option<Hash<32>>) -> bool {
```

Comment:
It is controversial, but in the Rust standard library `cmp` returns `Ordering`, so I think `hash_eq` would be better.

Reply:
Agree. better still just `impl PartialEq<Hash<32>>` instead of this custom named function.